lib/using_restricted_array.rb

```diff
@@ -1,61 +1,125 @@
+# Space complexity: O(1) because the values stored (i) do not increase in size with different array lengths
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0
+  
+  while i < length
+    return true if array[i] == value_to_find
+    i += 1
+  end
+  
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because the amount of time the method takes to complete in the worst case is  linear in relation to the length of the array
+# Space complexity: O(1) because the values stored (i and largest_int) do not increase in size with different array lengths
 def find_largest(array, length)
-  raise NotImplementedError
+  i = 0
+  largest_int = array[0]
+  
+  while i < length
+    largest_int = array[i] if array[i] > largest_int
+    i += 1
+  end
+  
+  return largest_int
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because the amount of time the method takes to complete in the worst case is linear in relation to the length of the array
+# Space complexity: O(1) because the values stored (i and smallest_int) do not increase in size with different array lengths
 def find_smallest(array, length)
-  raise NotImplementedError
+  i = 0
+  smallest_int = array[0]
+  
+  while i < length
+    smallest_int = array[i] if array[i] < smallest_int
+    i += 1
+  end
+  
+  return smallest_int
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because the amount of time the method takes to complete in the worst case is linear in relation to length of the array
+# Space complexity: O(1) because the values stored (j, i, a, and b) do not increase in size with different array lengths
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+  j = length - 1
+  while i < ( length / 2 )
+    a = array[i]
+    b = array[j]
+    
+    array[i] = b
+    array[j] = a
+    
+    i += 1
+    j -= 1
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log n) because the amount of time the method takes to complete in the worst case is logarithmic in relation to length of the array
+# Space complexity: O(1) because the values stored (j, i, mid) do not increase in size with different array lengths
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0
+  mid = length / 2
+  j = length - 1
+  
+  while i < j
+    return true if array[mid] == value_to_find || array[i] == value_to_find || array[j] == value_to_find
```

Comment:
While this works you could just compare `array[mid] == value_to_find` and drop the others.